gms/gossiper.cc

```diff
@@ -437,13 +437,30 @@ future<> gossiper::handle_echo_msg(gms::inet_address from, std::optional<int64_t
     return make_ready_future<>();
 }
 
-future<> gossiper::handle_shutdown_msg(inet_address from) {
+future<> gossiper::handle_shutdown_msg(inet_address from, std::optional<int64_t> generation_number_opt) {
     if (!is_enabled()) {
         logger.debug("Ignoring shutdown message from {} because gossip is disabled", from);
         return make_ready_future<>();
     }
-    return seastar::async([this, from] {
+    return seastar::async([this, from, generation_number_opt] {
         auto permit = this->lock_endpoint(from).get0();
+        if (generation_number_opt) {
+            auto es = this->get_endpoint_state_for_endpoint_ptr(from);
+            if (es) {
+                int local_generation = es->get_heart_beat_state().get_generation();
+                logger.info("Got shutdown message from {}, received_generation={}, local_generation={}",
+                        from, generation_number_opt.value(), local_generation);
+                if (local_generation != generation_number_opt.value()) {
```

Comment:
If the receiver doesn't yet know about the new generation of a node which went up but is now going down, it will incorrectly ignore the shutdown message. Maybe, if the generation is higher, we should still record it.

Reply:
It is totally fine to ignore the shutdown message. The receiver of the shutdown message can either learn from the other node which receives the shutdown message by gossip propagation or detect the node is down by the failure detector. There is no need to record it.

Reply:
Yes, it's not a serious problem to ignore it. But we send it for a reason right? If we think it's valuable to send it, it's better if we're reliable in delivering it.

Reply:
We send the shutdown message as an optimization to speed up other nodes knowing the shutdown of the node. However, as I mentioned, the node ignores the shutdown message will learn from gossip propagation pretty quickly from other node. I see no justification to record the message with higher generation to complicate the handling of the message.

Reply:
If it's true that gossip propagation is pretty quick and we don't need the shutdown message, why have the shutdown message at all?

I think it's not to make gossip propagation a bit it quicker, but to avoid timeout failures due to requests being routed to the node which is going down.

Reply:
See the example below:

n1, n2, n3
n3 was shutdown by user
n3 sent shutdown msg to n2, n2 received the msg and knows n3 is in shutdown status
n3 sent shutdown msg to n1, but n1 ignored it for some reason
n1 will learn n3 is shutdown status from n2 (any nodes that got the msg in the clsuter) by gossip propagation (A)

Without shutdown msg, all nodes had to detect n3 thought failure detector (B)
With shutdown msg, n3 tell the cluster actively. 

Procedure B is slow while A is fast. 

Shutdown msg has nothing to do with make gossip propagation faster or slower.

Reply:
This is also my point. So, we should make sure that it is reliably delivered, and we should care about the scenario which I brought up, right?

Reply:
Are you referring to the scenario "If the receiver doesn't yet know about the new generation of a node which went up but is now going down, it will incorrectly ignore the shutdown message"? This is totally fine. 

A node uses a new generation only after it restarts, during restart, it will send shutdown message to the cluster. Before this node goes up and  then shutdown again, other nodes in the cluster are supposed to know this node is down.  Before nodes know the new generation of the node that goes up, the node will be seen as down by other nodes. In this case, it's fine to ignore the shutdown message. If nodes learn the new generation, shutdown message will be in play. 

So, again, I do not think we need to record the shutdown message.